File: native-plugins/in-app-purchases/resources/android/src/com/buff/inapppurchases/SubscriptionsFunctions.kt
package com.buff.inapppurchases

import android.content.Context
import android.icu.text.MeasureFormat
import android.icu.util.Measure
import android.icu.util.MeasureUnit
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeError
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import com.nativephp.mobile.utils.NativeActionCoordinator
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PackageType
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.PurchasesErrorCode
import com.revenuecat.purchases.awaitCustomerInfo
import com.revenuecat.purchases.getOfferingsWith
import com.revenuecat.purchases.logInWith
import com.revenuecat.purchases.models.Period
import com.revenuecat.purchases.models.PricingPhase
import com.revenuecat.purchases.purchaseWith
import com.revenuecat.purchases.restorePurchasesWith
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.json.JSONObject
import java.util.Locale
import java.util.UUID

private object SubscriptionEvent {
    const val OFFERING_LOADED = "Buff\\InAppPurchases\\Events\\OfferingLoaded"
    const val OFFERING_FAILED = "Buff\\InAppPurchases\\Events\\OfferingFailed"
    const val PURCHASE_COMPLETED = "Buff\\InAppPurchases\\Events\\PurchaseCompleted"
    const val PURCHASE_CANCELLED = "Buff\\InAppPurchases\\Events\\PurchaseCancelled"
    const val PURCHASE_PENDING = "Buff\\InAppPurchases\\Events\\PurchasePending"
    const val PURCHASE_FAILED = "Buff\\InAppPurchases\\Events\\PurchaseFailed"
    const val RESTORE_COMPLETED = "Buff\\InAppPurchases\\Events\\RestoreCompleted"
    const val RESTORE_FAILED = "Buff\\InAppPurchases\\Events\\RestoreFailed"
}

private object SubscriptionState {
    val packages = mutableMapOf<String, Package>()
}

object SubscriptionsFunctions {
    class Configure(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val apiKey = parameters["api_key"] as? String
            if (apiKey == null || !Regex("^(goog|test)_[A-Za-z0-9]+$").matches(apiKey)) {
                throw BridgeError.InvalidParameters("api_key must be an Android or Test Store public SDK key")
            }

            val appUserID = parameters["app_user_id"] as? String
            if (appUserID == null || runCatching { UUID.fromString(appUserID) }.isFailure) {
                throw BridgeError.InvalidParameters("app_user_id must be a UUID")
            }

            if (!Purchases.isConfigured) {
                Purchases.configure(
                    PurchasesConfiguration.Builder(context, apiKey)
                        .appUserID(appUserID)
                        .build(),
                )

                return BridgeResponse.success(mapOf("configured" to true, "switching_account" to false))
            }

            if (Purchases.sharedInstance.appUserID == appUserID) {
                return BridgeResponse.success(mapOf("configured" to true, "switching_account" to false))
            }

            CoroutineScope(Dispatchers.Main).launch {
                Purchases.sharedInstance.logInWith(
                    appUserID = appUserID,
                    onSuccess = { _, _ -> },
                    onError = { },
                )
            }

            return BridgeResponse.success(mapOf("configured" to true, "switching_account" to true))
        }
    }

    class LoadOffering(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            if (!Purchases.isConfigured) {
                SubscriptionPayload.dispatch(
                    activity,
                    SubscriptionEvent.OFFERING_FAILED,
                    mapOf("category" to "not_configured", "message" to "Subscriptions are not configured."),
                )

                return BridgeResponse.success(mapOf("started" to false))
            }

            CoroutineScope(Dispatchers.Main).launch {
                Purchases.sharedInstance.getOfferingsWith(
                    onSuccess = { offerings ->
                        val packages = offerings.current?.availablePackages
                            ?.filter { it.packageType == PackageType.MONTHLY || it.packageType == PackageType.ANNUAL }
                            .orEmpty()

                        if (packages.isEmpty()) {
                            SubscriptionPayload.dispatch(
                                activity,
                                SubscriptionEvent.OFFERING_FAILED,
                                mapOf("category" to "unavailable", "message" to "No subscription offering is available."),
                            )
                            return@getOfferingsWith
                        }

                        SubscriptionState.packages.clear()
                        SubscriptionState.packages.putAll(packages.associateBy { it.identifier })
                        SubscriptionPayload.dispatch(
                            activity,
                            SubscriptionEvent.OFFERING_LOADED,
                            mapOf("packages" to packages.map(SubscriptionPayload::packagePayload)),
                        )
                    },
                    onError = { error ->
                        SubscriptionPayload.dispatch(
                            activity,
                            SubscriptionEvent.OFFERING_FAILED,
                            mapOf("category" to "offerings", "message" to error.message),
                        )
                    },
                )
            }

            return BridgeResponse.success(mapOf("started" to true))
        }
    }

    class Purchase(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val packageIdentifier = parameters["package_identifier"] as? String
            if (packageIdentifier.isNullOrBlank()) {
                throw BridgeError.InvalidParameters("package_identifier is required")
            }
            if (!Purchases.isConfigured) {
                throw BridgeError.ExecutionFailed("Subscriptions are not configured")
            }

            CoroutineScope(Dispatchers.Main).launch {
                val packageToPurchase = SubscriptionState.packages[packageIdentifier]
                if (packageToPurchase == null) {
                    SubscriptionPayload.dispatch(
                        activity,
                        SubscriptionEvent.PURCHASE_FAILED,
                        mapOf(
                            "category" to "unavailable",
                            "message" to "The selected subscription is unavailable.",
                            "package_identifier" to packageIdentifier,
                        ),
                    )
                    return@launch
                }

                Purchases.sharedInstance.purchaseWith(
                    purchaseParams = PurchaseParams.Builder(activity, packageToPurchase).build(),
                    onSuccess = { _, customerInfo ->
                        SubscriptionPayload.dispatch(
                            activity,
                            SubscriptionEvent.PURCHASE_COMPLETED,
                            mapOf(
                                "package_identifier" to packageIdentifier,
                                "product_identifier" to SubscriptionPayload.productIdentifier(packageToPurchase),
                                "entitled" to SubscriptionPayload.isEntitled(customerInfo),
                            ),
                        )
                    },
                    onError = { error, userCancelled ->
                        when {
                            userCancelled || error.code == PurchasesErrorCode.PurchaseCancelledError -> {
                                SubscriptionPayload.dispatch(
                                    activity,
                                    SubscriptionEvent.PURCHASE_CANCELLED,
                                    mapOf("package_identifier" to packageIdentifier, "category" to "cancelled"),
                                )
                            }
                            error.code == PurchasesErrorCode.PaymentPendingError -> {
                                SubscriptionPayload.dispatch(
                                    activity,
                                    SubscriptionEvent.PURCHASE_PENDING,
                                    mapOf(
                                        "package_identifier" to packageIdentifier,
                                        "product_identifier" to SubscriptionPayload.productIdentifier(packageToPurchase),
                                        "category" to "pending",
                                    ),
                                )
                            }
                            else -> {
                                SubscriptionPayload.dispatch(
                                    activity,
                                    SubscriptionEvent.PURCHASE_FAILED,
                                    mapOf(
                                        "category" to "purchase",
                                        "message" to error.message,
                                        "package_identifier" to packageIdentifier,
                                    ),
                                )
                            }
                        }
                    },
                )
            }

            return BridgeResponse.success(mapOf("started" to true))
        }
    }

    class Restore(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            if (!Purchases.isConfigured) {
                throw BridgeError.ExecutionFailed("Subscriptions are not configured")
            }

            CoroutineScope(Dispatchers.Main).launch {
                Purchases.sharedInstance.restorePurchasesWith(
                    onSuccess = { customerInfo ->
                        SubscriptionPayload.dispatch(
                            activity,
                            SubscriptionEvent.RESTORE_COMPLETED,
                            mapOf("entitled" to SubscriptionPayload.isEntitled(customerInfo)),
                        )
                    },
                    onError = { error ->
                        SubscriptionPayload.dispatch(
                            activity,
                            SubscriptionEvent.RESTORE_FAILED,
                            mapOf("category" to "restore", "message" to error.message),
                        )
                    },
                )
            }

            return BridgeResponse.success(mapOf("started" to true))
        }
    }

    class CustomerInfo(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            if (!Purchases.isConfigured) {
                return BridgeResponse.success(mapOf("configured" to false, "entitled" to false))
            }

            return try {
                val customerInfo = runBlocking(Dispatchers.IO) {
                    Purchases.sharedInstance.awaitCustomerInfo()
                }
                BridgeResponse.success(
                    mapOf("configured" to true, "entitled" to SubscriptionPayload.isEntitled(customerInfo)),
                )
            } catch (error: Exception) {
                throw BridgeError.ExecutionFailed(error.message ?: "Customer information is unavailable")
            }
        }
    }
}

private object SubscriptionPayload {
    fun dispatch(activity: FragmentActivity, event: String, payload: Map<String, Any>) {
        CoroutineScope(Dispatchers.Main).launch {
            NativeActionCoordinator.dispatchEvent(activity, event, JSONObject(payload).toString())
        }
    }

    fun isEntitled(customerInfo: CustomerInfo): Boolean =
        customerInfo.entitlements.active.containsKey("ai_meal_analysis")

    fun productIdentifier(packageToPurchase: Package): String =
        packageToPurchase.product.id.substringBefore(":")

    fun packagePayload(packageToPurchase: Package): Map<String, Any> {
        val product = packageToPurchase.product
        val payload = mutableMapOf<String, Any>(
            "package_identifier" to packageToPurchase.identifier,
            "product_identifier" to productIdentifier(packageToPurchase),
            "localized_price" to product.price.formatted,
            "localized_period" to localizedPeriod(product.period ?: product.defaultOption?.billingPeriod),
        )
        val introductoryPhase = product.defaultOption?.freePhase ?: product.defaultOption?.introPhase

        payload["introductory_offer"] = introductoryPhase?.let(::introductoryPayload) ?: JSONObject.NULL

        return payload
    }

    private fun introductoryPayload(phase: PricingPhase): Map<String, Any> = mapOf(
        "localized_price" to phase.price.formatted,
        "localized_period" to localizedPeriod(phase.billingPeriod),
        "period_count" to (phase.billingCycleCount ?: 1),
        "payment_mode" to (phase.offerPaymentMode?.name?.lowercase(Locale.ROOT) ?: "unknown"),
        "is_free_trial" to (phase.price.amountMicros == 0L),
    )

    private fun localizedPeriod(period: Period?): String {
        if (period == null) {
            return ""
        }

        val unit = when (period.unit) {
            Period.Unit.DAY -> MeasureUnit.DAY
            Period.Unit.WEEK -> MeasureUnit.WEEK
            Period.Unit.MONTH -> MeasureUnit.MONTH
            Period.Unit.YEAR -> MeasureUnit.YEAR
            Period.Unit.UNKNOWN -> return ""
        }

        return MeasureFormat.getInstance(Locale.getDefault(), MeasureFormat.FormatWidth.WIDE)
            .format(Measure(period.value, unit))
    }
}

File: native-plugins/in-app-purchases/resources/ios/Sources/SubscriptionsFunctions.swift
import Foundation
import RevenueCat

private enum SubscriptionEvent {
    static let configurationCompleted = "Buff\\InAppPurchases\\Events\\ConfigurationCompleted"
    static let configurationFailed = "Buff\\InAppPurchases\\Events\\ConfigurationFailed"
    static let offeringLoaded = "Buff\\InAppPurchases\\Events\\OfferingLoaded"
    static let offeringFailed = "Buff\\InAppPurchases\\Events\\OfferingFailed"
    static let purchaseCompleted = "Buff\\InAppPurchases\\Events\\PurchaseCompleted"
    static let purchaseCancelled = "Buff\\InAppPurchases\\Events\\PurchaseCancelled"
    static let purchasePending = "Buff\\InAppPurchases\\Events\\PurchasePending"
    static let purchaseFailed = "Buff\\InAppPurchases\\Events\\PurchaseFailed"
    static let restoreCompleted = "Buff\\InAppPurchases\\Events\\RestoreCompleted"
    static let restoreFailed = "Buff\\InAppPurchases\\Events\\RestoreFailed"
}

@MainActor
private enum SubscriptionState {
    static var packages: [String: Package] = [:]
}

enum SubscriptionsFunctions {
    class Configure: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let apiKey = parameters["api_key"] as? String,
                  apiKey.range(of: #"^(appl|test)_[A-Za-z0-9]+$"#, options: .regularExpression) != nil else {
                throw BridgeError.invalidParameters("api_key must be an iOS or Test Store public SDK key")
            }

            guard let appUserID = parameters["app_user_id"] as? String,
                  UUID(uuidString: appUserID) != nil else {
                throw BridgeError.invalidParameters("app_user_id must be a UUID")
            }

            if !Purchases.isConfigured {
                Purchases.configure(withAPIKey: apiKey, appUserID: appUserID)

                return BridgeResponse.success(data: ["configured": true, "switching_account": false])
            }

            guard Purchases.shared.appUserID != appUserID else {
                return BridgeResponse.success(data: ["configured": true, "switching_account": false])
            }

            Task { @MainActor in
                do {
                    _ = try await Purchases.shared.logIn(appUserID)
                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.configurationCompleted,
                        ["app_user_id": appUserID]
                    )
                } catch {
                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.configurationFailed,
                        [
                            "app_user_id": appUserID,
                            "category": "identity",
                            "message": error.localizedDescription,
                        ]
                    )
                }
            }

            return BridgeResponse.success(data: ["started": true, "switching_account": true])
        }
    }

    class LoadOffering: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard Purchases.isConfigured else {
                SubscriptionPayload.dispatch(
                    SubscriptionEvent.offeringFailed,
                    ["category": "not_configured", "message": "Subscriptions are not configured."]
                )

                return BridgeResponse.success(data: ["started": false])
            }

            Task { @MainActor in
                do {
                    guard let offering = try await Purchases.shared.offerings().current else {
                        SubscriptionPayload.dispatch(
                            SubscriptionEvent.offeringFailed,
                            ["category": "unavailable", "message": "No subscription offering is available."]
                        )
                        return
                    }

                    let packages = offering.availablePackages.filter {
                        $0.packageType == .monthly || $0.packageType == .annual
                    }
                    SubscriptionState.packages = Dictionary(uniqueKeysWithValues: packages.map { ($0.identifier, $0) })
                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.offeringLoaded,
                        ["packages": packages.map(SubscriptionPayload.package)]
                    )
                } catch {
                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.offeringFailed,
                        ["category": "offerings", "message": error.localizedDescription]
                    )
                }
            }

            return BridgeResponse.success(data: ["started": true])
        }
    }

    class Purchase: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let packageIdentifier = parameters["package_identifier"] as? String,
                  !packageIdentifier.isEmpty else {
                throw BridgeError.invalidParameters("package_identifier is required")
            }

            guard Purchases.isConfigured else {
                throw BridgeError.executionFailed("Subscriptions are not configured")
            }

            Task { @MainActor in
                guard let package = SubscriptionState.packages[packageIdentifier] else {
                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.purchaseFailed,
                        [
                            "category": "unavailable",
                            "message": "The selected subscription is unavailable.",
                            "package_identifier": packageIdentifier,
                        ]
                    )
                    return
                }

                do {
                    let result = try await Purchases.shared.purchase(package: package)

                    if result.userCancelled {
                        SubscriptionPayload.dispatch(
                            SubscriptionEvent.purchaseCancelled,
                            ["package_identifier": packageIdentifier, "category": "cancelled"]
                        )
                        return
                    }

                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.purchaseCompleted,
                        [
                            "package_identifier": packageIdentifier,
                            "product_identifier": package.storeProduct.productIdentifier,
                            "entitled": SubscriptionPayload.isEntitled(result.customerInfo),
                        ]
                    )
                } catch {
                    let code = ErrorCode(rawValue: (error as NSError).code)

                    if code == .purchaseCancelledError {
                        SubscriptionPayload.dispatch(
                            SubscriptionEvent.purchaseCancelled,
                            ["package_identifier": packageIdentifier, "category": "cancelled"]
                        )
                    } else if code == .paymentPendingError {
                        SubscriptionPayload.dispatch(
                            SubscriptionEvent.purchasePending,
                            [
                                "package_identifier": packageIdentifier,
                                "product_identifier": package.storeProduct.productIdentifier,
                                "category": "pending",
                            ]
                        )
                    } else {
                        SubscriptionPayload.dispatch(
                            SubscriptionEvent.purchaseFailed,
                            [
                                "category": "purchase",
                                "message": error.localizedDescription,
                                "package_identifier": packageIdentifier,
                            ]
                        )
                    }
                }
            }

            return BridgeResponse.success(data: ["started": true])
        }
    }

    class Restore: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard Purchases.isConfigured else {
                throw BridgeError.executionFailed("Subscriptions are not configured")
            }

            Task { @MainActor in
                do {
                    let customerInfo = try await Purchases.shared.restorePurchases()
                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.restoreCompleted,
                        ["entitled": SubscriptionPayload.isEntitled(customerInfo)]
                    )
                } catch {
                    SubscriptionPayload.dispatch(
                        SubscriptionEvent.restoreFailed,
                        ["category": "restore", "message": error.localizedDescription]
                    )
                }
            }

            return BridgeResponse.success(data: ["started": true])
        }
    }

    class CustomerInfo: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard Purchases.isConfigured else {
                return BridgeResponse.success(data: ["configured": false, "entitled": false])
            }

            return BridgeResponse.success(data: [
                "configured": true,
                "entitled": Purchases.shared.cachedCustomerInfo.map(SubscriptionPayload.isEntitled) ?? false,
            ])
        }
    }
}

private enum SubscriptionPayload {
    static func dispatch(_ event: String, _ payload: [String: Any]) {
        DispatchQueue.main.async {
            LaravelBridge.shared.send?(event, payload)
        }
    }

    static func isEntitled(_ customerInfo: RevenueCat.CustomerInfo) -> Bool {
        customerInfo.entitlements.active["buff_plus"] != nil
    }

    static func package(_ package: Package) -> [String: Any] {
        let product = package.storeProduct
        var payload: [String: Any] = [
            "package_identifier": package.identifier,
            "product_identifier": product.productIdentifier,
            "localized_price": product.localizedPriceString,
            "localized_period": localizedPeriod(product.subscriptionPeriod) ?? "",
        ]

        if let discount = product.introductoryDiscount {
            payload["introductory_offer"] = [
                "localized_price": discount.localizedPriceString,
                "localized_period": localizedPeriod(discount.subscriptionPeriod) ?? "",
                "period_count": discount.numberOfPeriods,
                "payment_mode": paymentMode(discount.paymentMode),
                "is_free_trial": discount.paymentMode == .freeTrial,
            ]
        } else {
            payload["introductory_offer"] = NSNull()
        }

        return payload
    }

    private static func localizedPeriod(_ period: SubscriptionPeriod?) -> String? {
        guard let period else {
            return nil
        }

        var components = DateComponents()
        let unit: NSCalendar.Unit

        switch period.unit {
        case .day:
            components.day = period.value
            unit = .day
        case .week:
            components.weekOfMonth = period.value
            unit = .weekOfMonth
        case .month:
            components.month = period.value
            unit = .month
        case .year:
            components.year = period.value
            unit = .year
        @unknown default:
            return nil
        }

        let formatter = DateComponentsFormatter()
        formatter.allowedUnits = unit
        formatter.unitsStyle = .full

        return formatter.string(from: components)
    }

    private static func paymentMode(_ mode: StoreProductDiscount.PaymentMode) -> String {
        switch mode {
        case .payAsYouGo:
            return "pay_as_you_go"
        case .payUpFront:
            return "pay_up_front"
        case .freeTrial:
            return "free_trial"
        @unknown default:
            return "unknown"
        }
    }
}
